rosserial_server/include/rosserial_server/session.h

```diff
@@ -83,34 +107,51 @@ class Session
         = boost::bind(&Session::handle_log, this, _1);
     callbacks_[rosserial_msgs::TopicInfo::ID_TIME]
         = boost::bind(&Session::handle_time, this, _1);
-  }
 
-  virtual ~Session()
-  {
-    ROS_INFO("Ending session.");
+    active_ = true;
+    attempt_sync();
+    read_sync_header();
   }
 
-  Socket& socket()
+  void stop()
   {
-    return socket_;
+    // Abort any pending ROS callbacks.
+    ros_callback_queue_.clear();
+
+    // Reset the state of the session, dropping any publishers or subscribers
+    // we currently know about from this client.
+    callbacks_.clear();
+    subscribers_.clear();
+    publishers_.clear();
+    services_.clear();
+
+    // Close the socket.
+    socket_.close();
+    active_ = false;
   }
 
-  void start()
+  bool is_active()
   {
-    ROS_INFO("Starting session.");
-
-    attempt_sync();
-    read_sync_header();
+    return active_;
   }
 
-  enum Version {
-    PROTOCOL_UNKNOWN = 0,
-    PROTOCOL_VER1 = 1,
-    PROTOCOL_VER2 = 2,
-    PROTOCOL_MAX
-  };
-
 private:
+  /**
+   * Periodic function which handles calling ROS callbacks, executed on the same
+   * io_service thread to avoid a concurrency nightmare.
+   */
+  void ros_spin_timeout(const boost::system::error_code& error) {
+    ros_callback_queue_.callAvailable();
```

Comment:
This is the kernel of the change— periodically calling ROS callbacks on the ASIO thread, rather than from a separate one.